include/revng/Pipeline/LLVMContainer.h

```diff
@@ -0,0 +1,201 @@
+  llvm::SmallVector<std::pair<std::uint32_t, llvm::MDNode *>, 4> Metadata;
+
+  llvm::WeakTrackingVH Mapped = Map[From];
+  T &To = llvm::cast<T>(*Mapped);
+  From->getAllMetadata(Metadata);
+
+  for (const auto &[index, Node] : Metadata) {
+    auto *ToMap = llvm::dyn_cast<llvm::MDNode>(Node);
+    if (ToMap == nullptr)
+      continue;
+
+    auto MayeMappedNode = Map.getMappedMD(ToMap);
+    if (not MayeMappedNode.hasValue())
+      continue;
+
+    auto *ToSet = llvm::dyn_cast<llvm::MDNode>(*MayeMappedNode);
+    if (ToSet == nullptr)
+      continue;
+    To.setMetadata(index, ToSet);
+  }
+}
+
+/// Implementation that can be derived by anyone so that multiple identical
+/// llvmcontainers can exists so that inspector kinds do not pollute each other.
+template<typename MostDerived>
+class LLVMContainerBase : public InspectableContainer<MostDerived> {
+private:
+  LLVMContainerImpl Container;
+
+public:
+  ~LLVMContainerBase() override {}
+
+  LLVMContainerBase(Context &Ctx, std::unique_ptr<llvm::Module> M) :
+    InspectableContainer<MostDerived>(Ctx), Container(std::move(M)) {}
+
+  const llvm::Module &getModule() const { return Container.getModule(); }
+  llvm::Module &getModule() { return Container.getModule(); }
+
+  void mergeBackImpl(MostDerived &&Other) final {
+    Container.mergeBackDerived(Other.Container);
+  }
+
+  llvm::Error storeToDisk(llvm::StringRef Path) const final {
+    return Container.storeToDisk(Path);
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef Path) final {
+    return Container.loadFromDisk(Path);
+  }
+
+  std::unique_ptr<ContainerBase>
+  cloneFiltered(const TargetsList &Targets) const final {
+    using InspectorT = LLVMContainerInspector<MostDerived>;
+    auto ToClone = InspectorT::enumerateAllFunctions(Targets, self());
+    auto ToClonedNotOwned = InspectorT::enumerateNonOnwedFunctions(self());
+
+    const auto Filter = [&ToClone, &ToClonedNotOwned](const auto &GlobalSym) {
+      if (not llvm::isa<llvm::Function>(GlobalSym))
+        return true;
+
+      const auto &F = llvm::cast<llvm::Function>(GlobalSym);
+      return ToClone.count(F) != 0 or ToClonedNotOwned.count(F) != 0;
+    };
+
+    llvm::ValueToValueMapTy Map;
+    auto Module = llvm::CloneModule(Container.getModule(), Map, Filter);
+
+    for (const llvm::Function &Original : Container.getModule().functions()) {
+      if (Original.isDeclaration())
+        cloneMetadata<llvm::Function>(Map, &Original);
+    }
+
+    return std::make_unique<MostDerived>(*this->Ctx, std::move(Module));
+  }
+
+  template<typename... LLVMPipePassess>
+  static PipeWrapper
+  wrapLLVMPassess(std::string LLVMModuleName, LLVMPipePassess &&...P) {
+    return PipeWrapper(LLVMPipeBase<MostDerived>(std::move(P)...),
+                       { std::move(LLVMModuleName) });
+  }
+
+  const LLVMContainerImpl &getContainer() const { return Container; }
+
+  LLVMContainerImpl &getContainer() { return Container; }
```

Comment:
Why? Is this used anywhere?
I've a *container* class with a `getContainer` method, confusing. 
Also, this seems to leak an implementation detail given we're giving the user a reference to an `...Impl` class.